c7n/resources/org.py

```diff
@@ -0,0 +1,361 @@
+# SPDX-License-Identifier: Apache-2.0
+
+from concurrent.futures import as_completed
+import logging
+import os
+import threading
+
+from botocore.exceptions import ClientError
+
+from c7n.credentials import assumed_session
+from c7n.executor import MainThreadExecutor
+from c7n.filters import Filter
+from c7n.query import QueryResourceManager, TypeInfo
+from c7n.resources.aws import AWS
+from c7n.utils import local_session, type_schema
+
+
+log = logging.getLogger("custodian.org-accounts")
+
+
+ORG_ACCOUNT_SESSION_NAME = "CustodianOrgAccount"
+
+
+def org_augment_tags(manager, resources):
+    client = local_session(manager.session_factory).client('organizations')
+    for r in resources:
+        r["Tags"] = client.list_tags_for_resource(
+            ResourceId=r["Id"]).get("Tags", [])
+    return resources
+
+
+class OrgAccess:
+
+    org_session = None
+
+    def parse_access_role(self):
+        params = {}
+        for q in self.data.get("query", ()):
+            params.update(q)
+        org_access = {
+            k: v for k, v in params.items() if k in ("org-access-role",)
+        }
+        return org_access.get('org-access-role')
+
+    def get_org_session(self):
+        # so we have to do a three way dance
+        # cli role -> (optional org access role) -> target account role
+        #
+        # in lambda though if we have a member-role we're effectively
+        # already in the org root on the event.
+        #
+        if self.org_session:
+            return self.org_session
+        org_access_role = self.parse_access_role()
+        if org_access_role and not (
+            "LAMBDA_TASK_ROOT" in os.environ and self.data.get("mode", {}).get("member-role")
+        ):
+            self.org_session = assumed_session(
+                role_arn=org_access_role,
+                session_name=ORG_ACCOUNT_SESSION_NAME,
+                region=self.session_factory.region,
+                session=local_session(self.session_factory),
+            )
+        else:
+            self.org_session = local_session(self.session_factory)
+        return self.org_session
+
+
+
+@AWS.resources.register("org-policy")
+class OrgPolicy(QueryResourceManager, OrgAccess):
+    policy_types = (
+        'SERVICE_CONTROL_POLICY',
+        'TAG_POLICY',
+        'BACKUP_POLICY',
+        'AISERVICES_OPT_OUT_POLICY',
+    )
+
+    class resource_type(TypeInfo):
+        service = "organizations"
+        id = "Id"
+        name = "Name"
+        arn = "Arn"
+        arn_type = "policy"
+        enum_spec = ("list_policies", "Policies", None)
+        global_resource = True
+        permissions_augment = ("organizations:ListTagsForResource",)
+
+    def resources(self, query=None):
+        q = self.parse_query()
+        if query is not None:
+            q.update(query)
+        else:
+            query = q
+        return super().resources(query=query)
+
+    def augment(self, resources):
+        return org_augment_tags(self, resources)
```

Comment:
Running policies in a sandbox org, I get this error:

```
An error occurred (InvalidInputException) when calling the ListTagsForResource operation: You provided a value that does not match the required pattern.
```

Looks like this trips up AWS-managed policies, presumably because those policies have different naming conventions like `p-FullAWSAccess`.

But it works fine on customer-managed policies. If I switch out to `universal_augment()` locally it works fine for both (just returns an empty tag set for the AWS-managed policy).

```suggestion
        return universal_augment(self, resources)
```

I'm thinking/hoping that would take advantage of the org session hopping here, but I'm not sure. Switching this would avoid the need for a permission augment also.

<details>

<summary>Full traceback</summary>

```console
An error occurred (InvalidInputException) when calling the ListTagsForResource operation: You provided a value that does not match the required pattern.
  File "/home/aj/.asdf/installs/python/c7n-dev-3.11/lib/python3.11/site-packages/botocore/client.py", line 1009, in _make_api_call
    raise error_class(parsed_response, operation_name)
  File "/home/aj/.asdf/installs/python/c7n-dev-3.11/lib/python3.11/site-packages/botocore/client.py", line 553, in _api_call
    return self._make_api_call(operation_name, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/aj/code/cloud-custodian/c7n/resources/org.py", line 29, in org_augment_tags
    r["Tags"] = client.list_tags_for_resource(
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/aj/code/cloud-custodian/c7n/resources/org.py", line 100, in augment
    return org_augment_tags(self, resources)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/aj/code/cloud-custodian/c7n/query.py", line 542, in resources
    with self.ctx.tracer.subsegment('resource-augment'):
  File "/home/aj/code/cloud-custodian/c7n/output.py", line 101, in subsegment (Current frame)
    yield self
  File "/home/aj/.asdf/installs/python/3.11.5/lib/python3.11/contextlib.py", line 155, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/home/aj/code/cloud-custodian/c7n/query.py", line 542, in resources
    with self.ctx.tracer.subsegment('resource-augment'):
  File "/home/aj/code/cloud-custodian/c7n/resources/org.py", line 97, in resources
    return super().resources(query=query)
  File "/home/aj/code/cloud-custodian/c7n/policy.py", line 330, in run
    resources = self.policy.resource_manager.resources()
  File "/home/aj/code/cloud-custodian/c7n/policy.py", line 1353, in __call__
    resources = PullMode(self).run()
  File "/home/aj/code/cloud-custodian/c7n/commands.py", line 307, in run
    policy()
  File "/home/aj/code/cloud-custodian/c7n/commands.py", line 128, in _load_policies
    return f(options, list(policies))
  File "/home/aj/code/cloud-custodian/c7n/cli.py", line 363, in main
    command(config)
  File "/home/aj/code/cloud-custodian/c7n/cli.py", line 372, in <module>
    main()
  File "/home/aj/.asdf/installs/python/3.11.5/lib/python3.11/runpy.py", line 88, in _run_code
    exec(code, run_globals)
  File "/home/aj/.asdf/installs/python/3.11.5/lib/python3.11/runpy.py", line 198, in _run_module_as_main
    return _run_code(code, main_globals, None,
botocore.errorfactory.InvalidInputException: An error occurred (InvalidInputException) when calling the ListTagsForResource operation: You provided a value that does not match the required pattern.
```

</details>

Reply:
yeah I don't think resource group tagging supported organization resources when this was first implemented, switching out to use that across the board makes sense to get bulk retrieval.

Reply:
added a flight recording test for tagging.